File: src/Graph.py

import heapq
import itertools
from Node import Node



class Graph:

    def __init__(self):
        self.graph = self.create_graph()

    def create_graph(self) -> list:
        Revelle = Node("Revelle College", ["college"])
        Warren = Node("Warren College", ["college"])
        Muir = Node("Muir College", ["college"])
        ERC = Node("Eleanor Roosevelt College", ["college"])
        Marshall = Node("Marshall College", ["college"])
        Sixth = Node("Sixth College", ["college"])
        Seventh = Node("Seventh College", ["college"])
        Eighth = Node("Eighth College", ["college"])
        RIMAC = Node("RIMAC", ["Gym(any)"])
        MainGym = Node("Main Gym", ["Gym(any)"])
        CVGym = Node("Canyon View Center", ["Gym(any)"])
        CafeV = Node("Cafe Ventanas", ["Food(any)"])
        OVT = Node("Ocean View Terrace", ["Food(any)"])
        Pines = Node("Pines", ["Food(any)"])
        Degrees64 = Node("64 Degrees", ["Food(any)"])
        PriceCenter = Node("Price Center", ["Food(any)", "Study(any)"])
        Center = Node("Center Hall", ["hall"])
        York = Node("York Hall", ["hall"])
        Galbraith = Node("Galbraith Hall", ["hall", "Study(any)"])
        PepperCanyon = Node("Pepper Canyon  Hall", ["hall"])
        Rady = Node("Rady School", ["hall"])
        SocialScience = Node("Social Science Building", ["hall"])
        FranklinAntonio = Node("Franklin Antonio Hall", ["hall", "Study(any)", "Food(any)"])
        CognitiveScience = Node("Cognitive Science Building", ["hall"])
        Mandeville = Node("Mandeville Auditorium", ["hall"])
        Jacobs = Node("Jacobs School", ["hall"])
        WarrenLecture = Node("Warren Lecture Hall", ["hall"])
        Geisel = Node("Geisel", ["Study(any)"])
        SuperComp = Node("SuperComputer Center", ["Study(any)"])
        CSE = Node("CSE Building", ["Study(any)"])

        Seventh.adjacent = {Rady : 1, CafeV : 1}
        CafeV.adjacent = {ERC : 1, Seventh : 1}
        ERC.adjacent = {CafeV : 1, RIMAC : 1, Marshall : 1, Rady : 1}
        Rady.adjacent = {Seventh : 1, RIMAC : 1, ERC : 1}
        RIMAC.adjacent = {SuperComp : 1, ERC : 1, Rady : 1}
        SuperComp.adjacent = {RIMAC : 1, SocialScience : 1}
        Marshall.adjacent = {ERC : 1, OVT : 1, Sixth : 1, SocialScience : 1}
        OVT.adjacent = {Marshall : 0}
        SocialScience.adjacent = {Marshall : 1, SuperComp : 1}
        Sixth.adjacent = {Muir : 1, Marshall : 1, CognitiveScience : 1}
        CognitiveScience.adjacent = {Geisel : 1, Sixth : 1}
        Geisel.adjacent = {CognitiveScience : 1, PriceCenter : 1, Jacobs : 1, WarrenLecture : 1}
        Jacobs.adjacent = {Geisel : 1, CSE : 1, WarrenLecture : 1, FranklinAntonio : 1}
        WarrenLecture.adjacent = {Geisel : 1, PriceCenter : 1, Jacobs : 1, CSE : 1, CVGym : 1}
        PriceCenter.adjacent = {Center : 1, Geisel : 1, WarrenLecture : 1}
        Center.adjacent = {PriceCenter : 1, Mandeville : 1}
        CSE.adjacent = {Warren : 1, Jacobs : 1, WarrenLecture : 1, CVGym : 1}
        Warren.adjacent = {FranklinAntonio : 1, CSE : 1, CVGym : 1}
        FranklinAntonio.adjacent = {Warren : 1, Jacobs : 1}
        CVGym.adjacent = {Warren : 1, CSE : 1, WarrenLecture : 1}
        Muir.adjacent = {Pines:  0, Sixth : 1, Mandeville : 1, MainGym : 1, Revelle : 1}
        Pines.adjacent = {Muir : 0}
        MainGym.adjacent = {Degrees64: 1, Mandeville : 1, Muir : 1}
        Mandeville.adjacent = {Muir : 1, Center : 1}
        Degrees64.adjacent = {Revelle : 0}
        Revelle.adjacent = {Muir : 1, Galbraith : 1, York : 1, Eighth : 1, Degrees64 : 0}
        Galbraith.adjacent = {York : 1, Revelle : 1, Eighth : 1}
        York.adjacent = {Galbraith : 1, Revelle : 1}
        Eighth.adjacent = {Revelle : 1, Galbraith : 1}
   
        node_list = [CafeV, CVGym, Center, CognitiveScience, CSE, ERC, Eighth, FranklinAntonio, 
                Jacobs, Galbraith, Geisel, MainGym, Mandeville, Marshall, Muir, 
                OVT, PepperCanyon, Pines, PriceCenter, Rady, Revelle, RIMAC, 
                Seventh, Sixth, SocialScience, SuperComp, Warren, WarrenLecture, York,  Degrees64]
        
        graph = {}
        for node in node_list:
            graph[node.name] = {'type': node.node_type, 'adjacent' : node.adjacent}
        return graph


    def find_shortest_path(self, start_name, end_name):
        #print(f"Finding shortest path from {start_name} to {end_name}")
        distances = {node.name: float('infinity') for node in self.node_list}
        previous_nodes = {node.name: None for node in self.node_list}
        distances[start_name] = 0
        pq = [(0, start_name)]

        while pq:
            current_distance, current_node_name = heapq.heappop(pq)
            #print(f"Current node: {current_node_name}, distance: {current_distance}")
            current_node = next(node for node in self.node_list if node.name == current_node_name)

            if current_distance > distances[current_node_name]:
                continue

            for neighbor, weight in current_node.adjacent.items():
                distance = current_distance + weight
                #print(f"  Checking neighbor: {neighbor.name}, distance: {distance}")

                if distance < distances[neighbor.name]:
                    distances[neighbor.name] = distance
                    previous_nodes[neighbor.name] = current_node_name
                    heapq.heappush(pq, (distance, neighbor.name))

        path = []
        current_node_name = end_name
        while current_node_name is not None:
            path.append(current_node_name)
            current_node_name = previous_nodes[current_node_name]
        path.reverse()

        #print(f"Shortest path: {path}, distance: {distances[end_name]}")
        return path, distances[end_name]

    def find_closest_location(self, current_location_name, activity_type):
        #print(f"Finding closest location to {current_location_name} for activity {activity_type}")
        min_distance = float('infinity')
        closest_location = None
        path_to_closest = []

        for node in self.node_list:
            if activity_type in node.node_type and node.name != current_location_name:
                #print(f"  Checking node: {node.name} for activity type {activity_type}")
                path, distance = self.find_shortest_path(current_location_name, node.name)
                if distance < min_distance:
                    min_distance = distance
                    closest_location = node.name
                    path_to_closest = path

        #print(f"Closest location: {closest_location}, path: {path_to_closest}, distance: {min_distance}")
        return closest_location, path_to_closest, min_distance

    def optimize_schedule(self, start, schedule):
        #print(f"Optimizing schedule from {start} with schedule {schedule}")
        current_location = start
        total_distance = 0
        full_path = [start]

        for item in schedule:
            #print(f"Processing schedule item: {item}")

            #current_node = next((node for node in self.node_list if node.name == current_node.name), None)
            current_node = next((node for node in self.node_list if node.name == current_location), None)


            if item in ["Food(any)", "Gym(any)", "Study(any)"]:

                if item in current_node.node_type:
                    closest_location = current_location

                    path = [current_location]
                    distance = 0
                else:
                    closest_location, path, distance = self.find_closest_location(current_location, item)

                #print(f"current_location : {current_location}")
                #print(f"closest_location : {closest_location}")
                #print(f"path : {path}")
                #print(f"distance : {distance}")
                if closest_location != current_location:
                    total_distance += distance
                    current_location = closest_location
                    full_path += path[1:]
                else:
                    full_path.append(current_location)
                #print(f"path[1:] : {path[1:]}")
                #full_path += path[1:]  # Include intermediate stops
                #full_path.append(current_location)
                #print(f"full_path: {full_path}")
                # Always append the current location for each schedule item
                #full_path.append(current_location)
            else:
                path, distance = self.find_shortest_path(current_location, item)
                total_distance += distance
                current_location = item

                full_path += path[1:]  # Include intermediate stops

                #print(f"full_path: {full_path}")

        #print(f"Optimized path: {full_path}, total distance: {total_distance}")
        return total_distance, full_path
    

        #Seventh.add_adjacency(Rady, 1)
        #Seventh.add_adjacency(CafeV, 1)
        #CafeV.add_adjacency(ERC, 1)
        #CafeV.add_adjacency(Seventh, 1)
        #ERC.add_adjacency(CafeV, 1)
        #ERC.add_adjacency(RIMAC, 1)
        #ERC.add_adjacency(Marshall, 1)
        #ERC.add_adjacency(Rady, 1)
        #Rady.add_adjacency(Seventh, 1)
        #Rady.add_adjacency(RIMAC, 1)
        #Rady.add_adjacency(ERC, 1)
        #RIMAC.add_adjacency(SuperComp, 1)
        #RIMAC.add_adjacency(ERC, 1)
        #RIMAC.add_adjacency(Rady, 1)
        #SuperComp.add_adjacency(RIMAC, 1)
        #SuperComp.add_adjacency(SocialScience, 1)
        #Marshall.add_adjacency(ERC, 1)
        #Marshall.add_adjacency(OVT, 0)
        #Marshall.add_adjacency(Sixth, 1)
        #Marshall.add_adjacency(SocialScience, 1)
        #OVT.add_adjacency(Marshall, 0)
        #SocialScience.add_adjacency(Marshall, 1)
        #SocialScience.add_adjacency(SuperComp, 1)
        #Sixth.add_adjacency(Muir, 1)
        #Sixth.add_adjacency(Marshall, 1)
        #Sixth.add_adjacency(CognitiveScience, 1)
        #CognitiveScience.add_adjacency(Geisel, 1)
        #CognitiveScience.add_adjacency(Sixth, 1)
        #Geisel.add_adjacency(CognitiveScience, 1)
        #Geisel.add_adjacency(PriceCenter, 1)
        #Geisel.add_adjacency(Jacobs, 1)
        #Geisel.add_adjacency(WarrenLecture, 1)
        #WarrenLecture.add_adjacency(Geisel, 1)
        #WarrenLecture.add_adjacency(PriceCenter, 1)
        #WarrenLecture.add_adjacency(Jacobs, 1)
        #WarrenLecture.add_adjacency(CSE, 1)
        #WarrenLecture.add_adjacency(CVGym, 1)
        #Jacobs.add_adjacency(Geisel, 1)
        #Jacobs.add_adjacency(CSE, 1)
        #Jacobs.add_adjacency(WarrenLecture, 1)
        #Jacobs.add_adjacency(FranklinAntonio, 1)
        #Center.add_adjacency(PriceCenter, 1)
        #Center.add_adjacency(Mandeville, 1)
        #PriceCenter.add_adjacency(Center, 1)
        #PriceCenter.add_adjacency(Geisel, 1)
        #PriceCenter.add_adjacency(WarrenLecture, 1)
        #CSE.add_adjacency(Warren, 1)
        #CSE.add_adjacency(Jacobs, 1)
        #CSE.add_adjacency(WarrenLecture, 1)
        #CSE.add_adjacency(CVGym, 1)
        #Warren.add_adjacency(FranklinAntonio, 1)
        #Warren.add_adjacency(CSE, 1)
        #Warren.add_adjacency(CVGym, 1)
        #FranklinAntonio.add_adjacency(Warren,1)
        #FranklinAntonio.add_adjacency(Jacobs, 1)
        #CVGym.add_adjacency(Warren, 1)
        #CVGym.add_adjacency(CSE, 1)
        #CVGym.add_adjacency(WarrenLecture, 1)
        #Muir.add_adjacency(Pines, 0)
        #Muir.add_adjacency(Sixth, 1)
        #Muir.add_adjacency(Mandeville, 1)
        #Muir.add_adjacency(MainGym, 1)
        #Muir.add_adjacency(Revelle, 1)
        #Pines.add_adjacency(Muir, 0)
        #MainGym.add_adjacency(Degrees64, 1)
        #MainGym.add_adjacency(Mandeville, 1)
        #MainGym.add_adjacency(Muir, 1)
        #Mandeville.add_adjacency(Muir, 1)
        #Mandeville.add_adjacency(Center, 1)
        #Degrees64.add_adjacency(Revelle, 0)
        #Revelle.add_adjacency(Muir, 1)
        #Revelle.add_adjacency(Galbraith, 1)
        #Revelle.add_adjacency(York, 1)
        #Revelle.add_adjacency(Eighth, 1)
        #Revelle.add_adjacency(Degrees64, 0)
        #Galbraith.add_adjacency(York, 1)
        #Galbraith.add_adjacency(Revelle, 1)
        #Galbraith.add_adjacency(Eighth, 1)
        #York.add_adjacency(Galbraith, 1)
        #York.add_adjacency(Revelle, 1)
        #Eighth.add_adjacency(Revelle, 1)
        #Eighth.add_adjacency(Galbraith, 1)


if __name__ == '__main__':

    graph = Graph()

    start = 'Warren Lecture Hall'

    schedule = ['Seventh College']

    result = graph.optimize_schedule(start, schedule)

    print(f"result : {result}")

File: src/Graph.js
import Node from './Node.js'


class Graph {
    constructor(props) {
        this.nodesMap = this.createGraph();
    }

    createGraph() {
        let nodes = {
            Revelle: new Node("Revelle College", ["college"]),
            Warren: new Node("Warren College", ["college"]),
            Muir: new Node("Muir College", ["college"]),
            ERC: new Node("Eleanor Roosevelt College", ["college"]),
            Marshall: new Node("Marshall College", ["college"]),
            Sixth: new Node("Sixth College", ["college"]),
            Seventh: new Node("Seventh College", ["college"]),
            Eighth: new Node("Eighth College", ["college"]),
            RIMAC: new Node("RIMAC", ["Gym(any)"]),
            MainGym: new Node("Main Gym", ["Gym(any)"]),
            CVGym: new Node("Canyon View Center", ["Gym(any)"]),
            CafeV: new Node("Cafe Ventanas", ["Food(any)"]),
            OVT: new Node("Ocean View Terrace", ["Food(any)"]),
            Pines: new Node("Pines", ["Food(any)"]),
            Degrees64: new Node("64 Degrees", ["Food(any)"]),
            PriceCenter: new Node("Price Center", ["Food(any)", "Study(any)"]),
            Center: new Node("Center Hall", ["hall"]),
            York: new Node("York Hall", ["hall"]),
            Galbraith: new Node("Galbraith Hall", ["hall", "Study(any)"]),
            PepperCanyon: new Node("Pepper Canyon Hall", ["hall"]),
            Rady: new Node("Rady School", ["hall"]),
            SocialScience: new Node("Social Science Building", ["hall"]),
            FranklinAntonio: new Node("Franklin Antonio Hall", ["hall", "Study(any)", "Food(any)"]),
            CognitiveScience: new Node("Cognitive Science Building", ["hall"]),
            Mandeville: new Node("Mandeville Auditorium", ["hall"]),
            Jacobs: new Node("Jacobs School", ["hall"]),
            WarrenLecture: new Node("Warren Lecture Hall", ["hall"]),
            Geisel: new Node("Geisel", ["Study(any)"]),
            SuperComp: new Node("SuperComputer Center", ["Study(any)"]),
            CSE: new Node("CSE Building", ["Study(any)"])
        };

        nodes.Seventh.adjacent = new Map([[nodes.Rady, 1], [nodes.CafeV, 1]]);
        nodes.CafeV.adjacent = new Map([[nodes.ERC, 1], [nodes.Seventh, 1]]);
        nodes.ERC.adjacent = new Map([[nodes.CafeV, 1], [nodes.RIMAC, 1], [nodes.Marshall, 1], [nodes.Rady, 1]]);
        nodes.Rady.adjacent = new Map([[nodes.Seventh, 1], [nodes.RIMAC, 1], [nodes.ERC, 1]]);
        nodes.RIMAC.adjacent = new Map([[nodes.SuperComp, 1], [nodes.ERC, 1], [nodes.Rady, 1]]);
        nodes.SuperComp.adjacent = new Map([[nodes.RIMAC, 1], [nodes.SocialScience, 1]]);
        nodes.Marshall.adjacent = new Map([[nodes.ERC, 1], [nodes.OVT, 0], [nodes.Sixth, 1], [nodes.SocialScience, 1]]);
        nodes.OVT.adjacent = new Map([[nodes.Marshall, 0]]);
        nodes.SocialScience.adjacent = new Map([[nodes.Marshall, 1], [nodes.SuperComp, 1]]);
        nodes.Sixth.adjacent = new Map([[nodes.Muir, 1], [nodes.Marshall, 1], [nodes.CognitiveScience, 1]]);
        nodes.CognitiveScience.adjacent = new Map([[nodes.Geisel, 1], [nodes.Sixth, 1]]);
        nodes.Geisel.adjacent = new Map([[nodes.CognitiveScience, 1], [nodes.PriceCenter, 1], [nodes.Jacobs, 1], [nodes.WarrenLecture, 1]]);
        nodes.Jacobs.adjacent = new Map([[nodes.Geisel, 1], [nodes.CSE, 1], [nodes.WarrenLecture, 1], [nodes.FranklinAntonio, 1]]);
        nodes.WarrenLecture.adjacent = new Map([[nodes.Geisel, 1], [nodes.PriceCenter, 1], [nodes.Jacobs, 1], [nodes.CSE, 1], [nodes.CVGym, 1]]);
        nodes.PriceCenter.adjacent = new Map([[nodes.Center, 1], [nodes.Geisel, 1], [nodes.WarrenLecture, 1]]);
        nodes.Center.adjacent = new Map([[nodes.PriceCenter, 1], [nodes.Mandeville, 1]]);
        nodes.CSE.adjacent = new Map([[nodes.Warren, 1], [nodes.Jacobs, 1], [nodes.WarrenLecture, 1], [nodes.CVGym, 1]]);
        nodes.Warren.adjacent = new Map([[nodes.FranklinAntonio, 1], [nodes.CSE, 1], [nodes.CVGym, 1]]);
        nodes.FranklinAntonio.adjacent = new Map([[nodes.Warren, 1], [nodes.Jacobs, 1]]);
        nodes.CVGym.adjacent = new Map([[nodes.Warren, 1], [nodes.CSE, 1], [nodes.WarrenLecture, 1]]);
        nodes.Muir.adjacent = new Map([[nodes.Pines, 0], [nodes.Sixth, 1], [nodes.Mandeville, 1], 
            [nodes.MainGym, 1], [nodes.Revelle, 1]]);
        nodes.Pines.adjacent = new Map([[nodes.Muir, 0]]);
        nodes.MainGym.adjacent = new Map([[nodes.Degrees64, 1], [nodes.Mandeville, 1], [nodes.Muir, 1]]);
        nodes.Mandeville.adjacent = new Map([[nodes.Muir, 1], [nodes.Center, 1]]);
        nodes.Degrees64.adjacent = new Map([[nodes.Revelle, 0]]);
        nodes.Revelle.adjacent = new Map([[nodes.Muir, 1], [nodes.Galbraith, 1], [nodes.York, 1], [nodes.Eighth, 1], [nodes.Degrees64, 0]]);
        nodes.Galbraith.adjacent = new Map([[nodes.York, 1], [nodes.Revelle, 1], [nodes.Eighth, 1]]);
        nodes.York.adjacent = new Map([[nodes.Galbraith, 1], [nodes.Revelle, 1]]);
        nodes.Eighth.adjacent = new Map([[nodes.Revelle, 1], [nodes.Galbraith, 1]]);


        this.nodeList = [nodes.CafeV, nodes.CVGym, nodes.Center, nodes.CognitiveScience, 
            nodes.CSE, nodes.ERC, nodes.Eighth, nodes.FranklinAntonio, 
            nodes.Jacobs, nodes.Galbraith, nodes.Geisel, nodes.MainGym, nodes.Mandeville, 
            nodes.Marshall, nodes.Muir, nodes.OVT, nodes.PepperCanyon, nodes.Pines, 
            nodes.PriceCenter, nodes.Rady, nodes.Revelle, nodes.RIMAC, 
            nodes.Seventh, nodes.Sixth, nodes.SocialScience, nodes.SuperComp,
            nodes.Warren, nodes.WarrenLecture, nodes.York,  nodes.Degrees64]
 

        let nodesMap = new Map();

        Object.values(nodes).forEach(node => {
            nodesMap.set(node.name, {
                type: node.nodeType,
                adjacent: node.adjacent
            });
        });

        return nodesMap;
    }


    findShortestPath(startName, endName) {
        // console.log(`Finding shortest path from ${startName} to ${endName}`);
        let distances = {};
        let previousNodes = {};
        this.nodeList.forEach(node => {
            distances[node.name] = Infinity;
            previousNodes[node.name] = null;
        });
        distances[startName] = 0;
        let pq = [[0, startName]];

        //console.log("nodeList : ")
        //for (let i = 0; i < this.nodeList.length; i++) {
          //  console.log(this.nodeList[i]); // Prints each element in the array
        //}
    
        while (pq.length) {
            pq.sort((a, b) => a[0] - b[0]); // This emulates a priority queue in JavaScript.
            let [currentDistance, currentNodeName] = pq.shift();
            // console.log(`Current node: ${currentNodeName}, distance: ${currentDistance}`);
            //console.log("NodeList : " + this.nodeList);
            let currentNode = this.nodeList.find(node => node.name === currentNodeName);
            
            //console.log("currentNode : " + currentNode);
            if (currentDistance > distances[currentNodeName]) {
                continue;
            }
            
            currentNode.adjacent.forEach((weight, neighbor) => {
                let distance = currentDistance + weight;
                // console.log(`  Checking neighbor: ${neighbor.name}, distance: ${distance}`);
    
                if (distance < distances[neighbor.name]) {
                    distances[neighbor.name] = distance;
                    previousNodes[neighbor.name] = currentNodeName;
                    pq.push([distance, neighbor.name]);
                }
            });
        }
    
        let path = [];
        let currentNodeName = endName;
        while (currentNodeName !== null) {
            path.push(currentNodeName);
            currentNodeName = previousNodes[currentNodeName];
        }
        path.reverse();
    
        // console.log(`Shortest path: ${path}, distance: ${distances[endName]}`);
        return [path, distances[endName]];
    }
    
    findClosestLocation(currentLocationName, activityType) {
        // console.log(`Finding closest location to ${currentLocationName} for activity ${activityType}`);
        let minDistance = Infinity;
        let closestLocation = null;
        let pathToClosest = [];
    
        this.nodeList.forEach(node => {
            if (node.nodeType.includes(activityType) && node.name !== currentLocationName) {
                // console.log(`  Checking node: ${node.name} for activity type ${activityType}`);
                let [path, distance] = this.findShortestPath(currentLocationName, node.name);
                if (distance < minDistance) {
                    minDistance = distance;
                    closestLocation = node.name;
                    pathToClosest = path;
                }
            }
        });
    
        // console.log(`Closest location: ${closestLocation}, path: ${pathToClosest}, distance: ${minDistance}`);
        return [closestLocation, pathToClosest, minDistance];
    }
    
    optimizeSchedule(start, schedule) {
        // console.log(`Optimizing schedule from ${start} with schedule ${schedule}`);
        let currentLocation = start;
        let totalDistance = 0;
        let fullPath = [start];
    
        schedule.forEach(item => {
            // console.log(`Processing schedule item: ${item}`);
    
            let currentNode = this.nodeList.find(node => node.name === currentLocation);
    
            if (["Food(any)", "Gym(any)", "Study(any)"].includes(item)) {
    
                if (currentNode.nodeType.includes(item)) {
                    let closestLocation = currentLocation;
                    let path = [currentLocation];
                    let distance = 0;
                } else {
                    console.log("Being fed into findclosestLocation: " + currentLocation)
                    let [closestLocation, path, distance] = this.findClosestLocation(currentLocation, item);
    
                    // console.log(`current_location : ${currentLocation}`);
                    // console.log(`closest_location : ${closestLocation}`);
                    // console.log(`path : ${path}`);
                    // console.log(`distance : ${distance}`);
                    if (closestLocation !== currentLocation) {
                        totalDistance += distance;
                        currentLocation = closestLocation;
                        fullPath = fullPath.concat(path.slice(1));
                    }
                }
            } else {
                let [path, distance] = this.findShortestPath(currentLocation, item);
                totalDistance += distance;
                currentLocation = item;
    
                fullPath = fullPath.concat(path.slice(1));
            }
    
            // console.log(`full_path: ${fullPath}`);
        });
    
        // console.log(`Optimized path: ${fullPath}, total distance: ${totalDistance}`);
        return [totalDistance, fullPath];
    }
}

export default Graph;